### PR TITLE
Roll forward Eigen to 5a0ab9f to solve the compilation problem with CUDA 9.1

### PR DESCRIPTION
**Pull request by pooyadavoodi**
This resolves a known issue with Eigen and CUDA 9.1
- https://stackoverflow.com/questions/43113508/math-functions-hpp-not-found-when-using-cuda-with-eigen
- https://github.com/tensorflow/tensorflow/issues/15389
- https://github.com/tensorflow/tensorflow/issues/15140

I don't know how we usually chose which Eigen commit to move to. I moved to the earliest that solves the problem. You may want to move to a later one though.

### CODE REVIEW

**Comment by Yangqing** [comment]
Thanks @pooyadavoodi - usually we pick a commit that has a tag on it, but if the fix is not behind a release tag yet, using the earliest makes sense.

**Review by Yangqing — APPROVED** [review]
Pending build passing - this looks good.

**Comment by bddppq** [comment]
@pooyadavoodi Thanks for the fix. I was trying to address this in a separate PR #6698 . Bumping Eigen does resolve the include issue, however, there is another issue with CUDA 9 that the compiler complains about __hadd can have multiple resolution (see here: https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-windows-build/638/console). I then tried a temporal fix to Eigen https://github.com/bddppq/eigen-git-mirror-fix-__hadd/commit/8e925c3855fca5bdb720e12b7eff8bf32f9d962b, but then it fails with another compilation error on windows saying this is invalid base class https://github.com/pytorch/pytorch/blob/master/caffe2/operators/arg_ops.h#L15, plus a lot of our gradient tests failed. 
Do you have a successful CUDA9 + Eigen build? If I remember correctly, we used to not use any Eigen stuffs in cuda operators, until recently arg_ops introduced this.

**Comment by pooyadavoodi** [comment]
hmm, I haven't seen any problem with CUDA 9. What is arg_ops, if that's added in the last couple of weeks, we haven't probably tested that. Let me compile with CUDA 9.

**Comment by bddppq** [comment]
Just to clarify, it's CUDA 9.1 :-)

**Comment by pooyadavoodi** [comment]
oh I see, CUDA 9.1 works for me after merging this PR.

**Comment by pooyadavoodi** [comment]
Just tried CUDA 9.0 as well. It works without this PR.

**Comment by bddppq** [comment]
Ok so it means `third_party\eigen\eigen\src/Core/arch/CUDA/Half.h(158): error : more than one instance of overloaded function "__hadd" matches the argument list` is a windows-only compilation error?

**Comment by pooyadavoodi** [comment]
Right, I guess only Windows. Is it on CUDA 9.0 or only 9.1?

**Comment by yinghai** [comment]
Judging from the change of error messages, I think it's CUDA 9.1.

**Comment by bddppq** [comment]
let's get this in first as it's a legit fix from the Eigen side and figure out windows build in separate PR

**Comment by pooyadavoodi** [comment]
Have you tested with the top of tree of Eigen?

**Comment by bddppq** [comment]
@pooyadavoodi yep I tried switching the eigen submodule to the official mirror which has more recent commits, and also tried a patch https://github.com/bddppq/eigen-git-mirror-fix-__hadd/commit/8e925c3855fca5bdb720e12b7eff8bf32f9d962b on top of that.
